src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -324,55 +326,151 @@ public ProgrammingExercise setupProgrammingExercise(ProgrammingExercise programm
 
         // save to get the id required for the webhook
         programmingExercise = programmingExerciseRepository.save(programmingExercise);
 
         versionControlService.get().addWebHook(testsRepoUrl, ARTEMIS_BASE_URL + TEST_CASE_CHANGED_API_PATH + programmingExercise.getId(), "ArTEMiS Tests WebHook");
+
         return programmingExercise;
     }
 
     // Copy template and push, if no file is in the directory
     private void setupTemplateAndPush(Repository repository, Resource[] resources, String prefix, String templateName, ProgrammingExercise programmingExercise) throws Exception {
         if (gitService.listFiles(repository).size() == 0) { // Only copy template if repo is empty
-            fileService.copyResources(resources, prefix, repository.getLocalPath().toAbsolutePath().toString());
-            if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA) {
-                fileService.replaceVariablesInDirectoryName(repository.getLocalPath().toAbsolutePath().toString(), "${packageNameFolder}",
-                        programmingExercise.getPackageFolderName());
+            fileService.copyResources(resources, prefix, repository.getLocalPath().toAbsolutePath().toString(), true);
+            replacePlaceholders(programmingExercise, repository);
+            pushRepository(repository, templateName);
+        }
+    }
+
+    /**
+     * Set up the test repository. This method differentiates non sequential and sequential test repositories (more than 1 test job).
+     *
+     * @param repository
+     * @param resources
+     * @param prefix
+     * @param templateName
+     * @param programmingExercise
+     * @throws Exception
+     */
+    private void setupTestTemplateAndPush(Repository repository, Resource[] resources, String prefix, String templateName, ProgrammingExercise programmingExercise)
+            throws Exception {
+        if (gitService.listFiles(repository).size() == 0 && programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA) { // Only copy template if repo is empty
+            String templatePath = "classpath:templates/" + programmingExercise.getProgrammingLanguage().toString().toLowerCase() + "/test";
+
+            String projectTemplatePath = templatePath + "/projectTemplate/**/*.*";
+            String testUtilsPath = templatePath + "/testutils/**/*.*";
+
+            Resource[] testUtils = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResources(testUtilsPath);
+            Resource[] projectTemplate = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResources(projectTemplatePath);
+
+            Map<String, Boolean> sectionsMap = new HashMap<>();
+
+            fileService.copyResources(projectTemplate, prefix, repository.getLocalPath().toAbsolutePath().toString(), false);
+
+            if (!programmingExercise.getSequentialTestRuns()) {
```

Comment:
can we write == true? I know it's not necessary, but the getter does not sound like a boolean

Reply:
You are right, the getter was auto generated but sounds weird.
I'll rename it to `hasSequentialTestRuns`.